### PR TITLE
Replace PyCrypto with pyca/cryptography

### PR DESCRIPTION
**Pull request by reaperhulk**
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
vault, crypto components

##### ANSIBLE VERSION
current devel branch

##### SUMMARY
This PR replaces PyCrypto, a largely moribund project written in poorly vetted bespoke C (with multiple outstanding CVEs), with pyca/cryptography.

pyca/cryptography is implicitly a dependency in many cases through paramiko (2.0+) as well as the new openssl_publickey module, which requires pyOpenSSL 16.0+. Additionally, pyca/cryptography is already presented as an optional dependency for better performance with Vault.

This commit leverages cryptography's padding module, constant time comparisons, fork safety, and block cipher constructs to reduce the amount of code ansible needs to maintain. Otherwise, I have attempted to minimize refactoring to ease review. Future PRs could further simplify several things.

While the code changes are only moderate, this is a significant dependency change. cryptography leverages OpenSSL/LibreSSL, so users on Linux will need to have the development headers for OpenSSL to be able to successfully install. This is an increased burden over the existing PyCrypto dependency (which only requires a C compiler). There are tentative plans to offer a binary wheel using `manylinux1` in the future, but that is not yet available. On Mac and Windows, binary wheels ship today so users simply need to have an up-to-date pip. Of course, since many users are already doing this to obtain paramiko 2.0+ the burden isn't too high. cryptography is also packaged by every major linux distribution, further reducing the challenge.

If this PR is accepted a future PR could easily remove the optional pyOpenSSL dependency and directly utilize cryptography for the public key module.

Full disclosure: I am one of the authors of pyca/cryptography. I strongly believe PyCrypto should no longer be used, but I do have a horse in this race :smile:

### CODE REVIEW

**Comment by reaperhulk** [comment]
I am almost certain I've missed some places where pycrypto is listed as a dependency, so please let me know if you find some!

**Comment by tiran** [comment]
PyCrypto's PBKDF2 implementation suffers from multiple issues and is about 30 times (!) slower than a good and correct implementation based on OpenSSL. Please use either cryptography or Python's hashlib module for PBKDF2. Here are some old benchmarks, https://bitbucket.org/tiran/backports.pbkdf2

Full disclosure: I'm the author of Python's ```hashlib.pbkdf2_hmac``` and a contributor to cryptography.

**Comment by reaperhulk** [comment]
Current CI status:

* `TEST=other` is failing because on Python 2.6 cryptography raises a `DeprecationWarning` which goes to stderr.
* Several of the docker containers are failing because of missing dependencies. I've updated their dockerfiles, but it looks like CI doesn't build a new container so it's not pulling in the updates?
* There are several issues within CI that are being uncovered by this change. paramiko was not being installed in many cases despite it being an install_require, and at least one builder has a version of six that is too out-of-date (https://app.shippable.com/runs/5886a2943b653a0f00fb06a7/37/console). These items are being caught because cryptography imports setuptools via pkg_resources, which enforces requirements and version constraints (which were previously being ignored).

It seems like it might make sense to split this into two (or three) separate PRs. One to address the CI dependencies (resolving the paramiko and six issues), then the actual cryptography change. Potentially a third PR to try to capture the dockerfile/script changes that add the libffi and openssl dependencies as well if we need to merge the dockerfile changes before using them in CI.

**Comment by abadger** [comment]
Initial feedback (I'll need to find someone else to give security feedback), I think we might be able to default to cryptography and fallback to pycrypto.  pycrypto's ubiquity vs cryptography's relative obscurity means that we can't have code that only works with cryptography.

I'm not sure what to do with the setup.py requirements in that case... Maybe we have to keep pycrypto in requirements but could make an extra_requires for cryptography?

Regarding PBKDF2 specifically, I think that the hashlib code isn't available in python-2.6 so we can't depend on it.  IIRC, the vault code already uses cryptography for PBKDF2 if available and falls back to pycrypto if it's not there.  We could probably add code that looks for either hashlib.pbkdf2 or cryptography, and then fallsback to pycrypto if someone wants to make a PR for that.

**Comment by reaperhulk** [comment]
`cryptography` is not as broadly available as pycrypto, but it definitely isn't particularly obscure at this point. Download statistics for the past 30 days from the PyPI BigQuery data:

```
cryptography	2,946,810
pycrypto	1,657,485
```

Distribution packaging accounts for some of that differential, but `cryptography` is also packaged with newer distributions (including Ubuntu 15.04+ and CentOS 7).

All that said, there would definitely be pain (especially with users trying to use the latest version of ansible on older distributions like CentOS < 6.5), so I understand the concern.

If the project doesn't want to take on that challenge right now then just using cryptography preferentially would be a good start. Fully purging the pycrypto requirement could then be part of an official deprecation cycle.

**Comment by mattclay** [comment]
Updates to our Dockerfiles need to be done in a separate PR. Once merged, the docker images can be rebuilt for use with future CI runs. I recommend building the docker images locally using your proposed changes, then testing your code changes with those images. Once tests pass, submit the PR to update our Dockerfiles. Once merged, we can rebuild your main PR and tests should pass in CI.

**Comment by mattclay** [comment]
@reaperhulk Is there a way to disable the deprecation warning on python 2.6? I don't think we want that warning displayed every time we run ansible, ansible-playbook, ansible-doc, etc.

**Comment by reaperhulk** [comment]
If you make sure the first import looks like:

```
import warnings

with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    from cryptography.hazmat.backends import default_backend
```

That will suppress the `DeprecationWarning` when run on Python 2.6.

I'll work on moving the Dockerfile changes to a separate PR soon, thanks for the information!

**Comment by bcoca** [comment]
I'm not sure we need to deal with those warnings, I would be shocked that those running 2.6 will be using the new library vs the old one.

**Comment by mattclay** [comment]
bot_status

**Comment by reaperhulk** [comment]
As an update: I still intend to split this up as suggested, but am on vacation at the moment.

File: test/runner/requirements/sanity.txt [hunk 1]
```diff
@@ -1,5 +1,7 @@
+cryptography
 jinja2
 mock
+paramiko
```
Comment (nicoonoclaste) [inline test/runner/requirements/sanity.txt:4]: Why is `paramiko` a new requirement here?

File: test/runner/requirements/windows-integration.txt [hunk 2]
```diff
@@ -1,4 +1,6 @@
+cryptography
 jinja2
 junit-xml
+paramiko
```
Comment (nicoonoclaste) [inline test/runner/requirements/windows-integration.txt:4]: Why is `paramiko` a new requirement here?

**Review by nicoonoclaste — COMMENTED** [review]
I will have a more thorough look later, but this looks promising.
(Note that I'm not an Ansible dev, just interesting in contributing to Ansible.)

**Comment by alikins** [comment]
I have a branch of this rebased to devel at https://github.com/ansible/ansible/compare/devel...alikins:cryptography_pr_20566_rebase

@reaperhulk If the rebase looks right, I can push it to your branch

**Comment by nicoonoclaste** [comment]
@alikins Did you include the extra patch I sent as reaperhulk/ansible#1 ?

**Comment by reaperhulk** [comment]
@alikins it mostly looks fine, although `ciphertext = self.read_data(filename)` on line 408 looks like it might have incorrect indentation now?

**Comment by alikins** [comment]
I re-rebased devel...alikins:cryptography_pr_20566_rebase to latest devel

**Comment by alikins** [comment]
@nbraud I cherry-picked 9e16b9db275263b3ea8d1b124966fdebfc9ab271 into https://github.com/ansible/ansible/compare/devel...alikins:cryptography_pr_20566_rebase

**Comment by alikins** [comment]
@reaperhulk 'ciphertext = self.read_data()' indention fixed

And a now unneeded ref to check_prereqs() was removed.

**Comment by tedder** [comment]
I'm glad to see this is being addressed. I'm less concerned about performance and more concerned about compatability/security/up-to-dateity.

**Comment by krzysztof-magosa** [comment]
@tedder Performance is also important. Current implementation is terrible - I need to wait about 20 for decryption of vaults within my group_vars.

**Comment by pcahyna** [comment]
FYI I've backported the rebased PR to the stable-2.3 branch and 2.3.0.0. You can fin the backports at https://github.com/pcahyna/ansible/tree/stable-2.3-no-more-pycrypto and https://github.com/pcahyna/ansible/tree/v2.3.0.0-1-no-more-pycrypto respectively.

**Comment by alikins** [comment]
Related to this, I have a work in progress branch for cipher_plugins at https://github.com/ansible/ansible/pull/25084 . The goal with that branch is to support multiple implementations of a given vault cipher/format so we could provide implementations for PyCrypto and 'cryptography' and choose the 'best' at runtime if more than one is found.

**Comment by alikins** [comment]
@reaperhulk I added 'cryptography' based plugins to my cipher_plugins branch based on these.
That branch is https://github.com/ansible/ansible/pull/25084

That branch doesnt include all the dep/setup/test changes, included here however since adding them to cipher_plugins branch is more a proof of concept...

**Comment by abadger** [comment]
The plan is for this PR to be superceded by: https://github.com/ansible/ansible/issues/25560

**Comment by reaperhulk** [comment]
(Closing in favor of abadger's work)

**Pull request closed without merging**